### PR TITLE
Refactor inventory_unit to support translations for predefined types

### LINKED ISSUE #6741: Normalize inventory units
Add support to inventory_units for translation for predefined unit names.

Ideally this would include abbreviations, plurals, and possibly a brief description of each type.  Maybe include a list of synonyms (for importing).  Also, a InventoryUnitService would make this easy to use.

One of the key design decisions would be to NOT include quantities in the inventory unit types.  For instance, in the past 'liters' and 'kg' have been known inventory unit types.  The inventory item description ('text') would include such quantity information.  Inventory items that might have been given 'liter' unit types in the past should be given a unit type of "bottle" (for example) and the fact that the bottle is 1 liter would be indicated in the inventory item itself (in its description/text).  However, it appears that we will need to include the counts of items for bottles/boxes containing a count of doses/items.

The 'inventory_unit' table would be updated to add a 'code' field for a translation key.  If present, then the i18n utilities can be used to look up the translation for the unit label and abbreviation (and perhaps handle pluralization).  If 'code' is not present the unit the inventory unit would work as they have in the past.  If 'code' is present, the inventory unit would be considered predefined and the editor for inventory units would not be allowed to change it.

Proposed list of units (based on current usage at several sites and feedback from future users).  Notice that there is some duplication (eg, pills/pillule vs tablet/comprime) that will need to be resolved.  Items marked with an asterisk are currently in the default data set in BHIMA (server/models/bhima.sql):

| English | French | text_id | Old id |
|---------|---------|------|-------|
| ampoule*     | ampoulle (amp)          |  AMPOULE | 6  |
| bag*         | sache                   |  BAG | 7  |
| bale         | ballot                  | BALE |   |
| barrel      | baril / tonneau              | BARREL |   |
| bottle      | boutelle* (btl)          |  BOTTLE | 8  |
| box*         | boite (bte)             | BOX |  4 |
| b/2           | b/2 |  B2 |   |
| b/4           | b/4 |  B4 |   |
| b/5           | b/5 | B5 |   |
| b/6           | b/6 | B6 |   |
| b/10         | b/10 | B10 |   |
| b/20           | b/20 | B20 |   |
| b/25           | b/25 | B25 |   |
| b/30          | b/30 | B30 |   |
| b/50           | b/50 | B50 |   |
| b/60           | b/60 | B60 |   |
| b/100           | b/100 | B100 |   |
| b/120           | b/120 | B120 |   |
| b/200           | b/200 | B200 |   |
| b/250           | b/250 | B250 |   |
| b/500           | b/500 | B500 |   |
| b/1000           | b/1000 | B1000 |   |
| bucket       | seau                    | BUCKET |   |
| bundle       | ballot                  | BUNDLE |   |
| can          | bidon                   | CAN |   |
| capsule*     | capsule (cap)           | CAPSULE | 9  |
| carton       | carton                  | CARTON |   |
| copie(s)     | Explaire (expl)         | COPY |   |
| days         | Jours                   | DAY |   |
| gloves       | gants                   | GLOVES |   |
| injection    | injectable (inj)        | INJECTION |   |
| jar*         | pot                     | JAR | 11  |
| kit          | kit                     | KIT |   |
| lot*         | lot                     | LOT |  5 |
| package      | colis                   | PACKAGE |   |
| packet       | paquet (pqt)            | PACKET |   |
| pallet | pallete | PALET | 2 |
| pair         | paire                   | PAIR |   |
| piece*       | piéce (pce)             | PIECE | 13  |
| pill         | pillule*, comprimé (pil) | PILL | 3  |
| pouch       | sachet*                  | POUCH |  14 |
| ream         | rame (rm)               | REAM |   |
| roll         | rouleau                 | ROLL |   |
| roller       | roulots (rol?)          | ==> ROLE |   |
| sack         | sac, jute               | SACK |   |
| set          | jeu                     | SET |   |
| tablet*      | comprimes (tab)         | TABLET |  15 |
| treatment    | cure                    | TREATMENT |   |
| tube*        | tube*                    | TUBE |  16 |
| unit        | unité*                   | UNIT | 18  |
| vial*        | vial                    | VIAL |  17 |
| | | | |
| article      | article (art)           |  ==> PIECE |  1 |
| bottle      | flacon* (flc)            | ==> BOTTLE |  10 ==> 8 |
| bundle       | botte                   | ==> BUNDLE |   |
| | littre | ==> BOTTLE | 12 |

### PR DESCRIPTION
**Pull request by jmcameron**
Refactored inventory_unit to support translations for predefined types.  Added more useful predefined types and English and French translations.  This refactor also makes sure that the inventory_units are referred to as 'unit_type' everywhere in the front end (before it was 'unit' in some places and 'unit_type' in others).

Added a few more inventory_unit types than proposed in https://github.com/IMA-WorldHealth/bhima/issues/6741 to support common unit types used at our production sites.

Here is a list of the new Inventory Unit types:  [DEFAULT BHIMA Inventory Units 2022-08-16.pdf](https://github.com/IMA-WorldHealth/bhima/files/9352007/DEFAULT.BHIMA.Inventory.Units.2022-08-16.pdf)

OLD version with b/nnn inventory units: ~~[DEFAULT BHIMA Inventory Units 2022-08-15.pdf](https://github.com/IMA-WorldHealth/bhima/files/9349038/DEFAULT.BHIMA.Inventory.Units.2022-08-15.pdf)~~

Closes https://github.com/IMA-WorldHealth/bhima/issues/6741

**TESTING**
- Use bhima_test -- make sure to do:  yarn build:db
- Go to:  Inventory > Inventory Configuration
  - Look at the new values in the `Inventory Unit form` part (bottom right)
  - Note that predefined values cannot be edited or delete
  - Try adding a new value.  Note that it can be edited or deleted.
- Go to:  Inventory > Inventory Registry
  - Note the `Unit` column.  For bhima_test, these should all be 'unit' (or 'unité') -- except for code=MOT.HCRF250RX which should show the unit type as 'lot'.
  - Try changing the language to observing the effect
  - Edit one of the inventory items and select a different unit.  Save and then verify the change
- Check out any of the pages that use inventory units (eg, Shipments, Stock Exits, and documents for Stock Exits) to verify that inventory units are handled correctlu

**MIGRATION OF PRODUCTION SITES!**
I have verified the migration behavior on one production site (Vanga, which is at the previous version of Bhima).  There are site-specific migration scripts for all the other active production sites.   All but Vanga are untested, but should work.  **WARNING** 
  - It is **VERY IMPORTANT** to make a backup of the site's database before trying to upgrade.  There is a significant chance that the migration scripts may need to be updated before they can be applied successfully.
  - You must determine which version of BHIMA is running on the production site you are upgrading
  - Do 'yarn migrate' to generate the full migration script.  Recall that 'yarn migrate' creates a full migration script -- including the site-specific part in separate files for each site.
  - After doing 'yarn migrate', the resulting full migration script will need to edited to insert the migration scripts between the release version in effect for the site and the current release in the appropriate place.   
  - ALSO, you may need to move the shipment specific triggers (shipment_reference, shipment_document_map) to a point in the full migration script AFTER the shipment table has been created!   BEFORE applying the script, check the full migration script to see if this change is necessary.

### CODE REVIEW

**Review by lomamech — COMMENTED** [review]
It's really an excellent idea, that of wanting to allow the list of inventory units, but I realize that there are us units that have been added.

My concern is in relation to those with the following prefix b/10, b100, b12 etc... because we held a meeting and decided to only interfere with the drugs in relation to the smallest unit, we took the resolution to be able to modify the label of the products without taking into account their packaging

_**Salbutamol Sulfate, 4mg, Tab, 1000** will be registered as **Salbutamol Sulfate, 4mg** without potency specified b1000_

**Comment by jmcameron** [comment]
> My concern is in relation to those with the following prefix b/10, b100, b12 etc... because we held a meeting and decided to only interfere with the drugs in relation to the smallest unit, we took the resolution to be able to modify the label of the products without taking into account their packaging

I understand this desire.  However, SEMI will be shipping boxes of medications and will not want to deal with individual items.  So perhaps we should support the B/200 inventory units but understand that when stock is entered into a dispensing depot that they will need to be converted at that point.   Then we will need to provide the functions in BHIMA to simplify that.

**Comment by lomamech** [comment]
> > My concern is in relation to those with the following prefix b/10, b100, b12 etc... because we held a meeting and decided to only interfere with the drugs in relation to the smallest unit, we took the resolution to be able to modify the label of the products without taking into account their packaging
> 
> I understand this desire. However, SEMI will be shipping boxes of medications and will not want to deal with individual items. So perhaps we should support the B/200 inventory units but understand that when stock is entered into a dispensing depot that they will need to be converted at that point. Then we will need to provide the functions in BHIMA to simplify that.

I think it's really possible to be able to do it, but this will lead to profound changes in BHIMA, a restructuring which will be really very deep, and will pose a problem for all our Production Sites (Tshikaji, Vanga, Karawa...)

During the meeting, the Pharmacist had revealed that the way in which the PNAM codifies drugs poses many problems for them.

For instance
Amoxycillin, 500mg, Caps, 100, Vrac **(CODE)DORA_AMOX5C-_3**
Amoxycillin, 500mg, Caps, 1000, Vrac **(CODE)DORA_AMOX5C_0**

Because with their codification we will face two totally different products, and their CMM and other indicator will be managed in a totally different way.

While the inventory management indicators do not take into account b100 or b1000 packaging

**Comment by lomamech** [comment]
@jmcameron
After parsing the old and new elements of the table, I realized that the ids of some inventory_unit have changed, I think it will be necessary that we can have scripts that should allow the update in our various production sites. 

image

**Comment by lomamech** [comment]
@jmcameron 
I also noticed that when viewing the list in French, although the name is displayed in alphabetical order, the numbers are no longer ordered, it will be preferable not to display the **database ID** but only the index of each line

image

**Comment by lomamech** [comment]
I think I'm done with the review as well as the functional analysis unless @mbayopanda can give his point of view

**Comment by jmcameron** [comment]
> @jmcameron I also noticed that when viewing the list in French, although the name is displayed in alphabetical order, the numbers are no longer ordered, it will be preferable not to display the **database ID** but only the index of each line

Ok

**Comment by jmcameron** [comment]
> @jmcameron After parsing the old and new elements of the table, I realized that the ids of some inventory_unit have changed, I think it will be necessary that we can have scripts that should allow the update in our various production sites.

Yes, there are separate scripts for each production site in the migration folder that fix the number changes.   Note these site-specific files get appended to the full migration file when you do 'yarn migrate'.

**Comment by jmcameron** [comment]
> > @jmcameron I also noticed that when viewing the list in French, although the name is displayed in alphabetical order, the numbers are no longer ordered, it will be preferable not to display the **database ID** but only the index of each line
> 
> Ok

I can change this, but what is the purpose of the numbers, if they all change any time a new inventory unit is added?  Perhaps a better idea is just to remove that column since is not something that the user needs or will ever use.

**Comment by lomamech** [comment]
> I can change this, but what is the purpose of the numbers, if they all change any time a new inventory unit is added? Perhaps a better idea is just to remove that column since is not something that the user needs or will ever use.



> > > @jmcameron I also noticed that when viewing the list in French, although the name is displayed in alphabetical order, the numbers are no longer ordered, it will be preferable not to display the **database ID** but only the index of each line
> > 
> > 
> > Ok
> 
> I can change this, but what is the purpose of the numbers, if they all change any time a new inventory unit is added? Perhaps a better idea is just to remove that column since is not something that the user needs or will ever use.

We can remove them

**Comment by jmcameron** [comment]
> We can remove them

Very good.  I'm working on it now.  And updating all the migration scripts!

**Comment by jmcameron** [comment]
Ok, I've removed the b/nnn type inventory units, renumbered, and updated the migration scripts and regtests.   I've also updated the PDF with the predefined inventory units for reference.  This file may be useful during migrations.

File: client/src/modules/inventory/configuration/units/units.html [hunk 1]
```diff
@@ -41,15 +41,20 @@
       <tbody>
         <tr ng-repeat="unit in UnitsCtrl.unitList | filter: UnitsCtrl.searchValue | orderBy: 'text' track by unit.id">
           <td>{{ $index + 1 }}</td>
-          <td>{{ unit.abbr }}</td>
           <td>{{ unit.text }}</td>
+          <td>{{ unit.abbr }}</td>
```
Comment (lomamech) [inline client/src/modules/inventory/configuration/units/units.html:45]: Cool

**Review by lomamech — APPROVED** [review]
I only hope the data from inventory unit table and IMCK is not different from KARAWA, I think the Pull Request is ready to be merged

**Comment by lomamech** [comment]
@mbayopanda also think about doing the code review

**Comment by jmcameron** [comment]
> I only hope the data from inventory unit table and IMCK is not different from KARAWA, I think the Pull Request is ready to be merged

I was not able to set up the site-specific migration script for Karawa because there are no backups that I could use to see what they are now using for inventory_unit data.    When the time comes to Upgrade Karawa, it will be necessary to copy and adapt one of the site-specific migrations scripts.

File: client/src/modules/inventory/configuration/units/units.service.js [hunk 2]
```diff
@@ -2,10 +2,27 @@ angular.module('bhima.services')
...
+
+  service.getUnits = function getUnits() {
+    return service.read()
+      .then(units => {
+        units.forEach(u => {
+          if (u.token) {
+            u.predefined = true;
+            u.text = $translate.instant(`INVENTORY.UNITS.${u.token}.TEXT`);
+            u.abbr = $translate.instant(`INVENTORY.UNITS.${u.token}.ABBR`);
+          } else {
```
Comment (mbayopanda) [inline client/src/modules/inventory/configuration/units/units.service.js:19]: 👍🏽

**Review by mbayopanda — APPROVED** [review]
LGTM 👍🏽 

bors r+

**Pull request merged**